test/teststl.cpp

```diff
@@ -3844,6 +3845,81 @@ class TestStl : public TestFixture {
               true);
         ASSERT_EQUALS("[test.cpp:4]: (style) Consider using std::accumulate algorithm instead of a raw loop.\n", errout.str());
     }
+
+    void findInsert()
+    {
+        check("void f1(std::set<unsigned>& s, unsigned x) {\n"
+              "    if (s.find(x) == s.end()) {\n"
+              "        s.insert(x);\n"
+              "    }\n"
+              "}\n",
+              true);
+        ASSERT_EQUALS("[test.cpp:3]: (performance) Searching before insertion is not necessary.\n", errout.str());
+
+        check("void f2(std::map<unsigned, unsigned>& m, unsigned x) {\n"
```

Comment:
Do you know if this checker is applicable for all containers or just the STL ones? Can we warn here if `QMap` is used here instead of `std::map`?

Reply:
I think it might work for some cases, but it looks like `QMap` is not an [`AssociativeContainer`](https://en.cppreference.com/w/cpp/named_req/AssociativeContainer), so the semantics of some the functions are different.

Reply:
If that is important we can add this info in the library.

Reply:
As far as I see it is not important what order the elements are stored in.. but I guess the property you want to have is that a (key) value can only be stored once.

Reply:
> As far as I see it is not important what order the elements are stored in.

Well its not the order but the semantics of the function which could incorrectly diagnose the issue. For example, `insert` should take the `value_type`, which is not the case for `QMap`. Also, `insert` for `QMap` is similar to C++'s `insert_or_assign` not C++'s `insert`. This is subtle but it could make the diagnostic confusing.

We should have a way to annotate a class as C++ container or Qt container.

Reply:
ok it sounds like the checking has to be more specific than I thought. Then we can have some STL hardcoding.